src/test/ui/span/issue-39018.stderr

```diff
@@ -64,17 +64,13 @@ help: `to_owned()` can be used to create an owned `String` from a string referen
 LL |     let _ = a + &b;
    |             ^   ^^
 
-error[E0308]: mismatched types
-  --> $DIR/issue-39018.rs:29:17
+error[E0277]: cannot add `std::string::String` to `std::string::String`
+  --> $DIR/issue-39018.rs:29:15
    |
 LL |     let _ = a + b;
-   |                 ^
-   |                 |
-   |                 expected &str, found struct `std::string::String`
-   |                 help: consider borrowing here: `&b`
+   |               ^ no implementation for `std::string::String + std::string::String`
```

Comment:
This would be an unfortunate error message regression. It doesn't necessarily mean we can't add the impl, but we would definitely want to follow up with a dedicated diagnostic to show that the user should write `a + &b`.

Reply:
I talked to Ruster-a11y about this and suggested they ignore it for now as a better error message can still be added later. Or do you think we should have a better error message in this PR already? My reasoning was that we first have to decide on the whole "breakage" situation anyway. Implementing the error message beforehand might be a waste of time.

I could have a go at implementing the better error message, if Ruster-a11y does not want to/does not have the time to.

Reply:
It doesn't need to be in this PR but it would be good to have something underway, so that we don't end up with the worse message for years. The code that the original message is explaining exactly how to fix is one of the things that many beginners struggle with.

Reply:
Absolutely! I'd say that the new impl and the better error message should hit stable at the same time. So absolutely not "years" in between.